### PR TITLE
Make andb_prop transparent

### PR DESCRIPTION
**Pull request by JasonGross**
Also give it a slightly simpler proof.

This makes `decide equality` give more transparent proofs in some cases.

### CODE REVIEW

**Comment by ppedrot** [comment]
I don't think that setting proof-related theorems transparent is a good thing. This breaks abstraction and can result costly. If what you want to solve is rather that decide equality is crappy as hell, then we should definitely fix that instead. (I did have plans to do that at some point in time...)

**Comment by JasonGross** [comment]
What I want to solve is the problem that when I transport across a concrete proof of equality given by `decide equality`, reduction is blocked on `andb_prop` for no good reason.

**Comment by ppedrot** [comment]
I'm assuming you're doing HoTT stuff in that case, or at least HoTT-like. I don't think that the stdlib has been designed with this particular point of view. What is your exact usecase?

**Comment by herbelin** [comment]
I have no particular opinion on this PR (both PMP's and Jason's arguments are defendable to me), nor on PMP's plans on decide equality. I'd just like to add a comment on the opaque to transparent issue which is regularly raised, which e.g. Abhishek raised again last 15 Nov on coq-club, which led him to work on Marc's parametricity plugin to find a solution (automatic duplication of an opaque constant into a transparent one), though I don't know if he could eventually get something working.

My comment is that I made a prototype 4 years ago to add a new command to turn an opaque constant into transparent a posteriori. I stopped at this time because I blocked on questions with the subtyping of the modules (thanks to Pierre L. for pointing out the problem). It seems now to me that the problem found by Pierre could be solved by changing the way in which module subtyping is done and if ever someone find this direction interesting to be investigated further, I can communicate my code.

[On the technical side, the idea to fix the problem I had is that "check_subtypes M S" not only return the universe constraints but also an update of M in which if we have an x:=t:T opaque in S, then the corresponding field of M in both case x:=t':T opaque and case x:T are transformed into x:=t:T opaque, thus forgetting t'. With a change of semantics compared to the current situation in the case x:=t ':T transparent in M: we would have it to be an error if t' <> t, which is not the case today.

We would thus ensure a coherence between implementation and specification even under opacity, so that if we eventually reveal M.x, which means to reveal the t of the signature (since the external vision of a module is its type), then the field in the implementation, and particularly in the extraction, would remain well consistent with the specification.]

So, I'm adding this comment for the record, being conscious that this is about a more general problem than just the problem of `andb_prop`. In practice, I don't think that continuing my work would be a lot of work, maybe only a couple of hours of work. The question is more about evaluating what is the best way to make progresses on the opaque to transparent issue:
- turning `Qed`s into `Defined`s (but losing the abstraction given by opacity)?
- automatically duplicating opaque definitions to a transparent copy of them (as in Abhishek)?
- to allow revealing bodies a posteriori as I started to do (and in full awareness that abstraction is then lost)?

**Comment by JasonGross** [comment]
@ppedrot Some of it is HoTT-informed, but it's not really HoTT.  I'm dealing with a PHOAS representation of terms where the type codes have decide equality.  Because this is PHOAS, I need to prove in some places that my syntax trees are well-formed (don't do silly things like case on var).  For large syntax trees (of the size that we get when, e.g., reifying field multiplication for GF25519 for 32-bit integers), proving well-formedness in Ltac (`repeat constructor`) is prohibitively expensive, because the proof tree for the inductive well-formedness relation is quadratic in the size of the expression.  Therefore, I needed to write a reflective partial solver for well-formedness, which constructs a type with significantly smaller proofs (in particular, one which is linear in the number of var nodes and constant nodes, and constant in the number of other nodes).  Because our syntax has let-in nodes (the same would pop up with application nodes), the partial solver needs to be heterogeneous in type codes.  Because, at variable nodes, we need to be able to derive a proof of homogenous equality, we can't use naive heterogeneous equality of variable nodes when the type doesn't line up.  I could have constructed a special heterogeneous equality type that does the right thing, and asked for a proof of UIP for the type codes, but I instead decided to ask for a transparent equality decider on type codes (because decide equality has this issue with `andb_prop`, I actually asked for a transparent equality semi-decider, because those are easier to define by hand).  (I believe that this choice gives slightly smaller theorem types and proof bodies for the kennel to check, and that was what I was trying to minimize.  Since it worked, I didn't bother trying the other option.  If `decide equality` gave transparent proofs, this option I chose would have been even simpler.)  The code is here: https://github.com/mit-plv/fiat-crypto/blob/7454086cafde4988dc115b7fb4ca5c0994ec09a1/src/Reflection/WfReflectiveGen.v

In short, I want this proof to be transparent because Coq is too slow at handling big proof terms and big theorem types.

**Comment by aa755** [comment]
Among the 3 choices at the end of Hugo's email, here are some pros and cons that come to my mind:
2) creates duplicates,  which can be confusing. It has an advantage over 3, namely that if the original opaque definition changes, only the objects that mention (perhaps indirectly) the duplicated version may need to be updated. 
Is there any reason to prefer 1) over 3)?

Between 2) and 3), I prefer 3), but I can live with 2).
Regarding ease of implementation, 2) should be easy once template-coq is complete -- currently, it lacks reification and reflection of universes. Its "Quote Recursively" quotes a term and (recursively) the bodies of all the (even opaque) members of the environment that it depends on. One can then take the result, let bind the environment, and ask template-coq to reflect the result back. A smarter version can try to only let-bind the opaque parts. I may be overlooking some universe-related issues.

**Comment by JasonGross** [comment]
@herbelin What's the difference between your options (1) and (3)?

**Comment by herbelin** [comment]
@JasonGross: (1) is like you proposed for `andb_prop`, i.e. to set transparent in advance in the library what is liable to be needed for computing. (3) is to provide a command, say `Reveal andb_prop` which makes the body visible only in a second step when computing starts to be more important than preserving the abstraction (though this is then irreversible).

**Comment by JasonGross** [comment]
@maximedenes This should be marked needs_discussion, not in_progress, right?

**Comment by maximedenes** [comment]
From what I understand, the real need here is to have `decide equality` parameterized so that it can be made to generate transparent proofs. @ejgallego is working on it, so I'm closing this PR.

**Comment by JasonGross** [comment]
Indeed.  I will be happy to see the day when there's a version of `decide equality ` that produces transparent proofs.
